pkg/devfile/adapters/kubernetes/component/adapter.go

```diff
@@ -423,102 +427,117 @@ func (a Adapter) waitAndGetComponentPod(hideSpinner bool) (*corev1.Pod, error) {
-		buildRequired = false
-	} else if len(pushDevfileCommands) == 2 {
-		// if there are two commands, it is the optional build command and the mandatory run command, set buildRequired to true
-		buildRequired = true
-	} else {
-		return fmt.Errorf("error executing devfile commands - there should be at least 1 command or at most 2 commands, currently there are %v commands", len(pushDevfileCommands))
+	if len(pushDevfileCommands) == 0 {
+		return errors.New(fmt.Sprint("error executing devfile commands - there should be at least 1 command."))
 	}
 
-	for i := 0; i < len(pushDevfileCommands); i++ {
-		command := pushDevfileCommands[i]
+	type CommandNames struct {
+		defaultName string
+		adapterName string
+	}
 
-		// Exec the devBuild command if buildRequired is true
-		if (command.Name == string(common.DefaultDevfileBuildCommand) || command.Name == a.devfileBuildCmd) && buildRequired {
-			glog.V(3).Infof("Executing devfile command %v", command.Name)
+	commandOrder := []CommandNames{}
 
-			for _, action := range command.Actions {
-				// Change to the workdir and execute the command
-				var cmdArr []string
-				if action.Workdir != nil {
-					cmdArr = []string{"/bin/sh", "-c", "cd " + *action.Workdir + " && " + *action.Command}
-				} else {
-					cmdArr = []string{"/bin/sh", "-c", *action.Command}
-				}
+	if runInit {
+		commandOrder = append(commandOrder, CommandNames{defaultName: string(common.DefaultDevfileInitCommand), adapterName: a.devfileInitCmd})
+	}
+	commandOrder = append(
+		commandOrder,
+		CommandNames{defaultName: string(common.DefaultDevfileBuildCommand), adapterName: a.devfileBuildCmd},
+		CommandNames{defaultName: string(common.DefaultDevfileRunCommand), adapterName: a.devfileRunCmd},
+	)
 
-				if show {
-					s = log.SpinnerNoSpin("Executing " + command.Name + " command " + fmt.Sprintf("%q", *action.Command))
+	for i, currentCommand := range commandOrder {
+		for _, command := range pushDevfileCommands {
+			if command.Name == currentCommand.defaultName || command.Name == currentCommand.adapterName {
+				if i < len(commandOrder)-1 {
+					// Any exec command such as "Init" and "Build"
+					err := a.executeDevfileCommand(command, show, podName)
+					if err != nil {
+						return err
+					}
 				} else {
-					s = log.Spinner("Executing " + command.Name + " command " + fmt.Sprintf("%q", *action.Command))
-				}
-
-				defer s.End(false)
-
-				err = exec.ExecuteCommand(&a.Client, podName, *action.Component, cmdArr, show)
-				if err != nil {
-					s.End(false)
-					return err
+					// Last command is "Run"
+					glog.V(3).Infof("Executing devfile command %v", command.Name)
+
+					for _, action := range command.Actions {
+
+						// Check if the devfile run component containers have supervisord as the entrypoint.
+						// Start the supervisord if the odo component does not exist
+						if !componentExists {
+							err = a.InitRunContainerSupervisord(*action.Component, podName, containers)
```

Comment:
We should check that `*action.Component` isn't nil or else we'll get some race errors.

Reply:
Already being done during push by validate func https://github.com/openshift/odo/blob/master/pkg/devfile/adapters/common/command.go#L85

@EnriqueL8 tho, just confirm this by removing a command's component before doing odo push